acceptable-albatrosses/albatrosses_hub/forums/views.py

```diff
@@ -0,0 +1,338 @@
+    such as the total number of posts in each board.
+    """
+    context = {"boards": Board.objects.all(), "login": request.user.is_authenticated, "page": "forum"}
+    return render(request, "forums/forum.html", context)
+
+
+class PostListView(ListView):
+    """
+    Main page for the message board, displays all posts in the message
+    board.
+    """
+
+    model = Post
+    paginate_by = 5
+    template_name = "forums/board.html"
+    context_object_name = "posts"
+
+    def get_queryset(self):
+        """
+        Queries and return all posts from the same message board.
+        """
+        self.board = get_object_or_404(Board, name=self.kwargs["board"])
+        return Post.objects.filter(board=self.board, comment__isnull=True).order_by("-created_at")
+
+    def get_context_data(self, **kwargs):
+        """
+        Adds board information into context (might not need this).
+        """
+        context = super().get_context_data(**kwargs)
+        context['board'] = self.board
+        context["login"] = self.request.user.is_authenticated
+        context["page"] = "forum"
+
+        return context
+
+
+class PostDetailView(DetailView):
+    """
+    View of a single post and all the comments on the posts.
+    """
+
+    model = Post
+    paginate_by = 5
+
+    def get_context_data(self, **kwargs):
+        """
+        Adds all comments of post into context data.
+        """
+        context = super().get_context_data(**kwargs)
+        context['comments'] = Comment.objects.filter(
+            post=context['post']).order_by('-created_at')
+        context["login"] = self.request.user.is_authenticated
+        context["page"] = "forum"
+
+        return context
+
+
+class PostCreateView(LoginRequiredMixin, CreateView):
+    """
+    View of page when creating posts.
+    """
+
+    model = Post
+    fields = ["subject", "message"]
+
+    def form_valid(self, form):
+        """
+        Fills in fields not queried from user.
+        """
+        form.instance.created_by = self.request.user
+        form.instance.board_id = get_object_or_404(Board, name=self.kwargs["board"]).id
+        return super().form_valid(form)
+
+    def get_context_data(self, **kwargs):
+        """
+        Get context from logged in user status and current page.
+        """
+        context = super().get_context_data(**kwargs)
+        context["login"] = self.request.user.is_authenticated
+        context["page"] = "forum"
+
+        return context
+
+
+class PostUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
+    """
+    View of page when editing posts.
+    """
+
+    model = Post
+    fields = ["subject", "message"]
+
+    def test_func(self):
+        """
+        To check user has permission (user is created_by) to edit post.
+        """
+        post = self.get_object()
+        if self.request.user == post.created_by:
+            return True
+        return False
```

Comment:
Some of the `Post` and `Comment` views share this exact method (minus a difference of `post` versus `comment` in the naming). At this point, it would be a good idea to factor this out. You could insert it as a mixin or using a decorator.